### PR TITLE
Add ability to disable partitions caching in hive metastore

### PR DESCRIPTION
**Pull request by s2lomon**
Add ability to disable partitions caching in hive metastore

    Thanks to this, one can disable caching of partitions, that
    should fix issues with missing partitions during query
    execution.

## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

It's a new feature that helps fixing some issues with inconsistencies in hive metastore caching

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

It's a change to a connector

> How would you describe this change to a non-technical end user or system administrator?

It allows you to pick whether you want to cache ALL, DOWN TO TABLE, or PARTITION metadata for hive metastore.

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java [hunk 1]
```diff
@@ -81,4 +84,31 @@ public CachingHiveMetastoreConfig setMaxMetastoreRefreshThreads(int maxMetastore
...
+
+    public MetastoreCacheScope getMetastoreCacheScope()
+    {
+        return metastoreCacheScope;
+    }
+
+    public enum MetastoreCacheScope
+    {
+        PARTITION,
+        TABLE,
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java:103]: Do you need TABLE now?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 2]
```diff
@@ -183,6 +213,23 @@ protected CachingHiveMetastore(HiveMetastore delegate, OptionalLong expiresAfter
         configValuesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, this::loadConfigValue);
     }
 
+    private static Predicate<MetastoreCacheScope> in(MetastoreCacheScope first, MetastoreCacheScope... others)
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:216]: add a method to enum instead

```
scope.matches(otherScope);
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 3]
```diff
@@ -137,42 +143,66 @@ public static CachingHiveMetastore memoizeMetastore(HiveMetastore delegate, long
...
 
-    protected CachingHiveMetastore(HiveMetastore delegate, OptionalLong expiresAfterWriteMillis, OptionalLong refreshMills, Optional<Executor> executor, long maximumSize, StatsRecording statsRecording)
+    protected CachingHiveMetastore(HiveMetastore delegate, MetastoreCacheScope metastoreCacheScope, OptionalLong expiresAfterWriteMillis, OptionalLong refreshMills, Optional<Executor> executor, long maximumSize, StatsRecording statsRecording)
     {
         this.delegate = requireNonNull(delegate, "delegate is null");
         requireNonNull(executor, "executor is null");
+        requireNonNull(metastoreCacheScope, "metastoreCacheScope is null");
 
-        databaseNamesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, ignored -> loadAllDatabases());
+        databaseNamesCache = metastoreCacheScope.whenMatching(in(ALL, TABLE),
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:160]: I would enable this only when ALL

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 4]
```diff
@@ -137,42 +143,66 @@ public static CachingHiveMetastore memoizeMetastore(HiveMetastore delegate, long
...
+                () -> buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, ignored -> loadAllDatabases()),
+                () -> neverCache(executor, ignored -> loadAllDatabases()));
 
-        databaseCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, this::loadDatabase);
+        databaseCache = metastoreCacheScope.whenMatching(in(ALL, TABLE),
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:164]: I would enable this only when ALL

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 5]
```diff
@@ -137,42 +143,66 @@ public static CachingHiveMetastore memoizeMetastore(HiveMetastore delegate, long
...
+                () -> neverCache(this::loadPartitionByName, this::loadPartitionsByNames));
 
-        tablePrivilegesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, key ->
-                loadTablePrivileges(key.getDatabase(), key.getTable(), key.getOwner(), key.getPrincipal()));
+        tablePrivilegesCache = metastoreCacheScope.whenMatching(in(ALL, TABLE),
+                () -> buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording,
+                        key -> loadTablePrivileges(key.getDatabase(), key.getTable(), key.getOwner(), key.getPrincipal())),
+                () -> neverCache(executor, key -> loadTablePrivileges(key.getDatabase(), key.getTable(), key.getOwner(), key.getPrincipal())));
 
         rolesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, ignored -> loadRoles());
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:207]: I would enable this only when ALL, same for below

**Review by kokosing — COMMENTED** [review]
Missing tests. But please

**Comment by Praveen2112** [comment]
> What is the use-case scenario when a user would want/need to use this?

It allows user to partially discard a cache - like we would like to cache the table details but not the partition information (if the user adds/drops a partition frequently)

**Comment by findepi** [comment]
> > What is the use-case scenario when a user would want/need to use this?
> 
> It allows user to partially discard a cache - like we would like to cache the table details but not the partition information (if the user adds/drops a partition frequently)

I know what the change does on a technical level.

It does not answer my question about a use-case scenario when a user would want/need to use this.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 6]
```diff
@@ -159,20 +160,35 @@ protected CachingHiveMetastore(HiveMetastore delegate, OptionalLong expiresAfter
...
+            partitionCache = buildCache(expiresAfterWriteMillis, maximumSize, statsRecording, this::loadPartitionByName, this::loadPartitionsByNames);
+        }
+        else {
+            partitionCache = neverCache(this::loadPartitionByName, this::loadPartitionsByNames);
+        }
 
-        tablePrivilegesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, key ->
-                loadTablePrivileges(key.getDatabase(), key.getTable(), key.getOwner(), key.getPrincipal()));
+        tablePrivilegesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording,
+                key -> loadTablePrivileges(key.getDatabase(), key.getTable(), key.getOwner(), key.getPrincipal()));
```
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:191]: Is this change required as a part of this commit ?

**Comment by Praveen2112** [comment]
> It does not answer my question about a use-case scenario when a user would want/need to use this.

In case of new partitions being added or dropped at some irregular interval - would make the system unusable at times at it fails (if partitions are dropped). We need to either disable the cache fully - but we might hit the underlying metastore too many times. This would be kind of a sweet spot between both the worlds.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java [hunk 7]
```diff
@@ -660,14 +712,77 @@ private static void await(CountDownLatch latch, long timeout, TimeUnit unit)
         }
     }
 
+    static class PartitionCachingAssertions
+    {
+        private final CachingHiveMetastore cachingHiveMetastore;
+        private final MockThriftMetastoreClient thriftClient;
+        private Consumer<CachingHiveMetastore> sutAction = hiveMetastore -> {};
```
Comment (Praveen2112) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:719]: Is there is any expansion for `sutAction` ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java [hunk 8]
```diff
@@ -81,4 +82,16 @@ public CachingHiveMetastoreConfig setMaxMetastoreRefreshThreads(int maxMetastore
         this.maxMetastoreRefreshThreads = maxMetastoreRefreshThreads;
         return this;
     }
+
+    @Config("hive.metastore-partition-cache-enabled")
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java:86]: `hive.metastore-cache-partitions`

however, we should have `hive.metastore-cache.*` prefix, so maybe  `hive.metastore-cache.cache-partitions`  and rename existing configs in a follow-up.

cc @losipiuk @electrum

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java [hunk 9]
```diff
@@ -81,4 +82,16 @@ public CachingHiveMetastoreConfig setMaxMetastoreRefreshThreads(int maxMetastore
...
+    public CachingHiveMetastoreConfig setPartitionCacheEnabled(boolean enabled)
+    {
+        this.partitionCacheEnabled = enabled;
+        return this;
+    }
+
+    public boolean isPartitionCacheEnabled()
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java:93]: put getter above the setter

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastoreConfig.java [hunk 10]
```diff
@@ -33,6 +33,7 @@ public void testDefaults()
                 .setMetastoreCacheTtl(new Duration(0, TimeUnit.SECONDS))
                 .setMetastoreRefreshInterval(null)
                 .setMetastoreCacheMaximumSize(10000)
+                .setPartitionCacheEnabled(true)
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastoreConfig.java:36]: You chose to add the field as the last one in the config class, so add it as a last one in the test class too.
(or move declaration in the config class)

(same below)

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 11]
```diff
@@ -119,7 +119,7 @@
     private final LoadingCache<String, Set<RoleGrant>> grantedPrincipalsCache;
     private final LoadingCache<String, Optional<String>> configValuesCache;
 
-    public static CachingHiveMetastore cachingHiveMetastore(HiveMetastore delegate, Executor executor, Duration cacheTtl, Optional<Duration> refreshInterval, long maximumSize)
+    public static CachingHiveMetastore cachingHiveMetastore(HiveMetastore delegate, Executor executor, Duration cacheTtl, Optional<Duration> refreshInterval, long maximumSize, boolean partitionCacheEnabled)
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:122]: Sure thing.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 12]
```diff
@@ -141,10 +142,10 @@ public static CachingHiveMetastore memoizeMetastore(HiveMetastore delegate, long
                 OptionalLong.empty(),
                 Optional.empty(),
                 maximumSize,
-                StatsRecording.DISABLED);
+                StatsRecording.DISABLED, true);
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:145]: sure

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 6 above)
Comment (s2lomon) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:191]: nope, it's my mistake

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java (hunk 9 above)
Comment (s2lomon) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java:93]: sure

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java [hunk 13]
```diff
@@ -799,7 +799,8 @@ protected final void setup(String host, int port, String databaseName, String ti
                 executor,
                 new Duration(1, MINUTES),
                 Optional.of(new Duration(15, SECONDS)),
-                10000);
+                10000,
+                true);
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:803]: ok

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java (hunk 7 above)
Comment (s2lomon) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:719]: sut is a shortcut for `system under test`. It's a way to mark this as a holder for all actions that are actually being tested. I could rename it to `cachingHiveMetastoreInteractions` or `metastoreInteractions/Actions`. For me sut is ok, but maybe the fact that you need to know what it means makes it worst than one of the above.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastoreConfig.java (hunk 10 above)
Comment (s2lomon) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastoreConfig.java:36]: ok

**Comment by s2lomon** [comment]
> > It does not answer my question about a use-case scenario when a user would want/need to use this.
> 
> In case of new partitions being added or dropped at some irregular interval - would make the system unusable at times at it fails (if partitions are dropped). We need to either disable the cache fully - but we might hit the underlying metastore too many times. This would be kind of a sweet spot between both the worlds.

I guess it should fix issues like this one.  #6286 Or at least would be a good try for such scenarios. Alternatively we could add fine tuning parameters, but I guess it's easier to have it cached as always or not at all.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:803]: @s2lomon 
if this tests's purpose is  to test things related to partitions being cached or not, then i agree, the config should be set explicitly.
however, it's not the goal here. you supplied the config only because the compiler ordered you to (missing parameter), not that because you really wanted to enable or disable partition caching.
`new ..Config().is...CacheaPartitions()` conveys this intent, and neither `true` nor `false`  conveys that intent

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 14]
```diff
@@ -183,6 +200,16 @@ protected CachingHiveMetastore(HiveMetastore delegate, OptionalLong expiresAfter
         configValuesCache = buildCache(expiresAfterWriteMillis, refreshMills, executor, maximumSize, statsRecording, this::loadConfigValue);
     }
 
+    private static <K, V> LoadingCache<K, V> neverCache(Optional<Executor> executor, com.google.common.base.Function<K, V> loader)
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:203]: the `executor` (refresh executor) is useless for a "never cache", drop the param

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java (hunk 8 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java:91]: i suggest `hive.metastore-cache.cache-partitions` and rename existing configs in a follow-up.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:803]: switch to `new ..Config().is...CacheaPartitions()` please

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java [hunk 15]
```diff
@@ -660,14 +705,78 @@ private static void await(CountDownLatch latch, long timeout, TimeUnit unit)
...
+        PartitionCachingAssertions whenExecuting(Consumer<CachingHiveMetastore> interactions)
+        {
+            this.metastoreInteractions = interactions;
+            return this;
+        }
+
+        void usesCache()
+        {
+            IntStream.range(1, 5)
+                    .forEach(times -> {
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:740]: do ordinary `for`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java [hunk 16]
```diff
@@ -660,14 +705,78 @@ private static void await(CountDownLatch latch, long timeout, TimeUnit unit)
...
+                        metastoreInteractions.accept(cachingHiveMetastore);
+                        assertEquals(thriftClient.getAccessCount(), 1, "Metastore is expected to use cache, but it does not.");
+                    });
+        }
+
+        void doesNotUseCache()
+        {
+            IntStream.range(1, 5)
+                    .forEach(times -> {
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java:749]: do ordinary `for`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastoreConfig.java [hunk 17]
```diff
@@ -44,12 +45,14 @@ public void testExplicitPropertyMappings()
...
                 .put("hive.metastore-cache-maximum-size", "5000")
                 .put("hive.metastore-refresh-max-threads", "2500")
+                .put("hive.metastore-partition-cache-enabled", "false")
                 .buildOrThrow();
 
         CachingHiveMetastoreConfig expected = new CachingHiveMetastoreConfig()
                 .setMetastoreCacheTtl(new Duration(2, TimeUnit.HOURS))
                 .setMetastoreRefreshInterval(new Duration(30, TimeUnit.MINUTES))
                 .setMetastoreCacheMaximumSize(5000)
+                .setPartitionCacheEnabled(false)
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastoreConfig.java:55]: make it last (as everywhere else)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 18]
```diff
@@ -1021,6 +1048,7 @@ private static <K, V> LoadingCache<K, V> buildCache(
         if (statsRecording == StatsRecording.ENABLED) {
             cacheBuilder.recordStats();
         }
+        cacheBuilder.shareNothingWhenDisabled();
```
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:1051]: Why do we need to invoke this now ?

**Review by Praveen2112 — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 18 above)
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java:1051]: Because we use EvictableCache and as of now configure some of them not to cache ("disable")

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (s2lomon) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java]: Omg you are right, for some reason I've thought that this comment was about the default test. Good point and thanks for being patient :)

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java (hunk 7 above)
Comment (findepi) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/cache/TestCachingHiveMetastore.java]: > sut is a shortcut for `system under test`.

would never guess

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java [hunk 19]
```diff
@@ -81,4 +82,16 @@ public CachingHiveMetastoreConfig setMaxMetastoreRefreshThreads(int maxMetastore
         this.maxMetastoreRefreshThreads = maxMetastoreRefreshThreads;
         return this;
     }
+
+    public boolean isPartitionCacheEnabled()
+    {
+        return partitionCacheEnabled;
+    }
+
+    @Config("hive.hive.metastore-cache.cache-partitions")
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreConfig.java:91]: to many hives here

**Review by findepi — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
CI https://github.com/trinodb/trino/issues/11275 (reopened)

**Comment by mosabua** [comment]
@colebow can you verify if we need docs and if yes .. raise a PR asap.
